draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1581,75 +1581,121 @@ Steps:
 26. return (x, y)
 ~~~
 
-## Mappings for twisted Edwards curves
+## Mappings for Edwards curves
 
-### Elligator 2 Method {#ell2edwards}
+Edwards curves are closely related to Montgomery
+curves ({{montgomery}}): every Edwards curve is birationally equivalent
+to a Montgomery curve ({{BBJLP08}}, Theorem 3.2).
+This equivalence yields an efficient way of hashing to an Edwards curve:
+first, hash to the equivalent Montgomery curve, then transform the
+result into a point on the Edwards curve by evaluating a pair of rational
+functions called a birational map.
+Thus, hashing to an Edwards curve requires first identifying a
+corresponding Montgomery curve and birational map.
+
+### Birational maps from Montgomery to Edwards curves {#birational-map}
+
+There are two ways to identify the Montgomery curve and birational map
+for use with a given Edwards curve.
 
-Twisted Edwards curves and Montgomery curves are closely related: every
-twisted Edwards curve is birationally equivalent to a Montgomery curve
-({{BBJLP08}}, Theorem 3.2). To hash to a twisted Edwards curve, hash
-to the equivalent Montgomery curve and evaluate the rational map to
-obtain a point on the twisted Edwards curve.
+When hashing to a standard Edwards curves for which a corresponding
+Montgomery form and birational map are also standardized, the standard
+Montgomery form and birational map MUST be used to ensure compatibility
+with existing software.
+Two examples of this are the edwards25519 and edwards448 curves,
+which correspond to the Montgomery curves curve25519 and curve448, respectively.
+For both of these curves, {{RFC7748}} lists both the Montgomery and Edwards
+forms and gives the corresponding rational maps.
```

Comment:
Should we note that `sqrt` in those implementations MUST be chosen from the set we recommend in this document? (That is, must be constant time.)

Reply:
This is a great catch.

Since those square roots are constants, I would guess that most people will just hard code a constant in their implementation.

But: the sign of the square root isn't specified, so the sign of the resulting point is ambiguous!

To me, the best way to fix this is to do what we've done everywhere else: specify that the sign of the Edwards point should be equal to the sign of the input `u`, since this is both unambiguous and insensitive to whether future standards manage to find other ways to ambiguously specify rational maps.

Reply:
That sounds good!

Reply:
Argh, as I'm thinking about this more, I'm not super happy with this solution. But I'm not sure whether there's any other solution that's better.

The reason I'm not super happy is, if we don't tell folks how to resolve the ambiguity in RFC7748, then it may still break things in annoying ways. Specifically, I'm worried about a case where someone wants to interoperate between curve25519 and edwards25519, and wants to be able to hash to both of them.

As currently written, there's no guarantee that a conforming hash-to-Montgomery and a conforming hash-to-Edwards will give the same result, because the maps could disagree on sign. It gets worse in the case of the random oracle construction: since the maps may disagree on sign sometimes and agree on sign other times, there are 4 possible outputs:

- P + Q
- P - Q
- -P + Q
- -P - Q

This is bad! I think we want the curve25519 and edwards25519 hash functions to give exactly the same result, and I think for that purpose the only thing we can do is specify the sign of the birational map.

I'm going to push a revert for the sign commit and push another commit with this solution. That way, we can pick and choose among them.

Reply:
To be sure I understand, the invariant you're after is:

```
phi(HM(x)) == HE(x)
```

where `HM`, `HE`, and `phi` are the hash-to-Montgomery, hash-to-Edwards, and Montgomery-to-Edwards functions, respectively. By asserting that each map uses positive constants, the goal is that `phi` would not change the point sign. Is that right? (I need to think through this more.)

Reply:
The goal of fixing the sign of the constants is just to make sure that everyone uses exactly the same definition of phi(), so that different implementations will interoperate. In the case of edwards25519, we'll have test vectors in the standard, so everything should be fine. The remaining concern is that future standardization efforts for Edwards curves will have the same ambiguity. Resolving ambiguously-signed constants in one direction *should* suffice to prevent this.

By the way, positive constants don't guarantee that the sign is unchanged, because "sign" acts a little weird here: the product of two positive numbers can be negative, for example. (As a simple example, consider the integers mod 13. By our definition, both 2 and 6 are positive, but their product is negative.)

Reply:
I was playing fast and loose with my words -- ensuring that everyone's phi implementation is the same is what I was after. I agree with the conclusion that fixing the sign ensures compatibility now and in the future.

Reply:
:+1: sorry to be pedantic, just wanted to be sure we were on the same page. :)

Reply:
```
phi(HM(x)) == HE(x)
```
I am not sure whether we can get this equation working at the hashing-to-curve level. T
he main reason is that both hashing methods are prefixed with different strings, which includes the name of curve in the prefix.

Reply:
> Making sure that the Edwards and Montgomery hashes always give equivalent results helps with this kind of compatibility, because it means that the Edwards PAKE library could use any conforming hash-to-Edwards implementation.

It is not clear to me, why to recommend two different hashes HE(x) and HM(x) that map to the "same" equivalent point (under certain birational mapping).

What we need is 1 hash that covers any Montgomery curve and 1 hash that covers any Twisted Edwards curve.

The relation between curve25519 and edwards25519 is different to 
the relation between curve448 and edwards448. These are special cases.

I do not understand why the hash to curve method must carry with these relations.

Reply:
> I am not sure whether we can get this equation working at the hashing-to-curve level. The main reason is that both hashing methods are prefixed with different strings, which includes the name of curve in the prefix.

This is true *only* if hash-to-curve includes a ciphersuite tag. Right now it seems like most of the opinions expressed in #124 are against having a ciphersuite tag at the hash-to-curve level, so it's at least reasonable to believe that this could be made to work. But it's a good point that it's not clear that we're "doing enough" here.

> It is not clear to me, why to recommend two different hashes HE(x) and HM(x) that map to the "same" equivalent point (under certain birational mapping).

The reason is that *in practice* people care about being able to use the Montgomery and Edwards forms of a curve interchangeably, so we should try to make sure that this standard accommodates this use-case.

For example, this issue came up as described in a comment in #107: initially the document suggested using a different birational map than the one in RFC7748, and we got pushback against this. So this is not just a hypothetical concern.

Reply:
FWIW, I was assuming we removed the ciphersuite string from this document, as @kwantam states above.